### DISCUSSION TITLE
Next Auth, SIWE, Rainbow Kit with Next JS 13 Route Handlers

### DISCUSSION THREAD

**Discussion by 0xRowdy** [nextauthjs/next-auth#7923] [Help] [answered]
### Question 💬

I'm trying to implement Next Auth with Sign in with Ethereum and Rainbow Kit using the new Next JS 13 Route Handlers.

I've got the functionality working but I'm receiving errors both in the browser and in the terminal.

Errors:
cmd line:
```typescript
https://next-auth.js.org/errors#client_fetch_error Unexpected token 'E', "Error: Thi"... is not valid JSON {
  error: {
    message: `Unexpected token 'E', "Error: Thi"... is not valid JSON`,
    stack: `SyntaxError: Unexpected token 'E', "Error: Thi"... is not valid JSON\n` +
      '    at JSON.parse (<anonymous>)\n' +
      '    at parseJSONFromBytes (node:internal/deps/undici/undici:6571:19)\n' +
      '    at successSteps (node:internal/deps/undici/undici:6545:27)\n' +
      '    at node:internal/deps/undici/undici:1211:60\n' +
      '    at node:internal/process/task_queues:140:7\n' +
      '    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)\n' +
      '    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)\n' +
      '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)',
    name: 'SyntaxError'
  },
  url: 'http://localhost:3000/api/auth/csrf',
  message: `Unexpected token 'E', "Error: Thi"... is not valid JSON`
}
```

Browser:
```
Object { context: "core" }
 
Object { context: "core/relayer" }
 Error: WebSocket connection closed abnormally with code: 3000 ```

Reference documentation:
https://www.rainbowkit.com/docs/authentication
https://docs.login.xyz/integrations/nextauth.js
https://wagmi.sh/examples/sign-in-with-ethereum
https://nextjs.org/docs/app/building-your-application/routing/router-handlers

I'm new to Next and Next auth so I may have not configured things correctly. If there is a repo or code sandbox example of this working I would be forever grateful.

Thanks in advance for any help.

### How to reproduce ☕️

I created a bare-bones repo to showcase the errors I am getting. Clone the repo, Install dependencies, and configure the .env.local. You should be able to connect the wallet, sign a message for next auth but will get errors in the browser console and terminal.

I tried setting up a code sandbox but was having very intermittent issues with it loading.

https://github.com/0xRowdy/nextauth-siwe-route-handlers/tree/main

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by walmat** [comment]
Did you end up finding a solution to this? I'm experiencing the same thing.

**Comment by MattStew87** [comment]
I am having the first problem whenever I click on the authenticate button it prints that to my terminal. But also even though my application works in development and production it is unable to authenticate at deployment time, citing errors with my credentialProvider. Lmk if anyone figures any part of this out.

**Comment by truemiller** [comment]
I am also experiencing this issue...

**Comment by 0xRowdy** [comment]
Still seems to be an issue. It is also an issue if you follow this tutorial using Web3 modal instead of Rainbow Kit.
https://dev.to/abhikbanerjee99/nextjs-13-how-to-web3-gg9

**Comment by luisalrp** [comment]
Hey @0xRowdy! Thanks for your work. Your example is pretty clean and helped me a lot. I had the same issue as you and I used the following code:
`const message = JSON.parse(credentials?.message)
const signature = credentials?.signature
const csrfToken = credentials?.csrfToken
const fields = await siweMessage.validate(signature)
console.log('fields: ', fields)
if (fields.nonce !== csrfToken) return new Response(JSON.stringify({ message: 'Invalid nonce.' }), { status: 422 })`

The token it returns in the jwt callback the first time is:
`token: {
    token: { sub: '0x4...1b14' },
    user: { id: '0x4...1b14' },
    account: {
      providerAccountId: '0x23.....14',
      type: 'credentials',
      provider: 'credentials'
    },
    isNewUser: false,
    trigger: 'signIn',
    iat: 1693333254,
    exp: 1695925254,
    jti: 'e643e8d9-4a79-47f1-bc89-6dd54c962dfc'
  },`

I would like to see the provider as "ethereum" or "SIWE" but I don't really know how to do that.

Hope it helps! Cheers!

**Comment by LuisJoseLopezD** [comment]
same issue

**Comment by KoenRijpstra** [comment, marked as answer]
You can fix this issue by changing `await getCsrfToken({ req })` to `await getCsrfToken({ req: { headers: req.headers } })`. 

It should look like this:
```js
const result = await siwe.verify({
  signature: credentials?.signature || "",
  domain: nextAuthUrl.host,
  nonce: await getCsrfToken({ req: { headers: req.headers } }),
});
```

More info here: https://github.com/nextauthjs/next-auth/issues/7166

**Reply by 0xRowdy** [reply]
Awesome. Thanks for posting that here!

**Comment by keennay** [comment]
> https://github.com/0xRowdy/nextauth-siwe-route-handlers/tree/main

Looks like I get this same issue as the OP using WalletConnect on Next 14:
https://github.com/nextauthjs/next-auth/discussions/10357

`getCsrfToken()` already has `{ req: { headers: req.headers } }` per below:
https://docs.walletconnect.com/web3modal/nextjs/siwe/next-auth

**Reply by keennay** [reply]
Got this quickly solved, seems multiple issues can all return an "Unexpected token" error.

In my case since I'm running this on a dedicated server for production use, I also needed the following environment variable alongside NEXTAUTH_URL:
NEXTAUTH_URL_INTERNAL

Example usage:
NEXTAUTH_URL_INTERNAL=http://127.0.0.1:3000
